docs/project_practice/how_to_build_a_todo_application_with_ribir.md

```diff
@@ -0,0 +1,464 @@
+
+We call Todo interface it `todo_widget`.
+
+```rust
+use ribir::prelude::*;
+
+fn main() {
+  // ...
+  let todo_widget = widget! {
+    Column {
+      Input { Placeholder::new("What do you want to do?") }
+      Lists {
+        ListItem {
+          Leading {
+            Checkbox { checked: false }
+          }
+          HeadlineText::new("Complete how to build an application with Ribir")
+        }
+      }
+    }
+  };
+  app::run(todo_widget);
+}
+```
+
+You can try execute it by `cargo run`, you will see a window what it show `Input` and `Lists`.
+
+## Combine of data and view
+
+Now we have Todo data and view, how we combine this both? Actually our data is core, view is just the presentation of data. So we need modify view code.
+
+Here we need use `Compose` trait to combine data and view. Let's try use `Compose` modify above code case.
+
+```rust
+// ...
+impl Compose for TodoMVP {
+  fn compose(this: StateWidget<Self>) -> Widget {
+    widget! {
+      states { this: this.into_stateful() }
+      Column {
+        Input { Placeholder::new("What do you want to do?") }
+        Lists {
+          ListItem {
+            Leading {
+              Checkbox { checked: false }
+            }
+            HeadlineText::new("Complete how to build an application with Ribir")
+          }
+        }
+      }
+    }
+  }
+}
+
+fn main() {
+  let todo_data = TodoMVP {
+    tasks: vec![
+      Task {
+        done: false,
+        description: "Complete how to build an application with Ribir".to_string(),
+      }
+    ]
+  }.into_stateful();
+  app::run(todo_data.into_widget());
+}
+```
+
+And task data not static, we need to render it dynamically, here we need `DynWidget` to represent dynamic widgets.
+
+```rust
+use ribir::prelude::*;
+
+#[derive(Clone)]
+struct Task {
+  done: bool,
+  description: String,
+}
+
+struct TodoMVP {
+  tasks: Vec<Task>,
+}
+
+impl TodoMVP {
+  fn task(task: Task) -> Widget {
+    widget! {
+      ListItem {
+        Leading {
+          Checkbox { checked: task.done }
+        }
+        HeadlineText::new(task.description.clone())
+      }
+    }
+  }
+}
+
+impl Compose for TodoMVP {
+  fn compose(this: StateWidget<Self>) -> Widget {
+    widget! {
+      states { this: this.into_stateful() }
+      Column {
+        Input { Placeholder::new("What do you want to do?") }
+        Lists {
+          DynWidget {
+            dyns: {
+              let tasks = this.tasks.clone();
+              tasks
+                .into_iter()
+                .map(move |task| {
+                  TodoMVP::task(task)
+                })
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+fn main() {
+  let todo_data = TodoMVP {
+    tasks: vec![
+      Task {
+        done: false,
+        description: "Complete how to build an application with Ribir".to_string(),
+      }
+    ]
+  }.into_stateful();
```

Comment:
Unnecessary `into_stateful()`

Reply:
We todo example use it, so todo example don't need it also?